File: citybackend/src/main/java/com/cityprojects/citybackend/service/profile/ProfileService.java
package com.cityprojects.citybackend.service.profile;

import com.cityprojects.citybackend.dto.profile.ChangePasswordDto;
import com.cityprojects.citybackend.dto.profile.ProfileDto;
import com.cityprojects.citybackend.dto.profile.ProfileUpdateDto;
import org.springframework.web.multipart.MultipartFile;

/**
 * Service de gestion du profil self-service.
 *
 * <h2>Pas de @RequireTenant</h2>
 * <p>Toutes les operations sont scopees a l'<b>utilisateur courant</b> resolu
 * via {@link com.cityprojects.citybackend.common.security.SecurityUtils#currentUserIdOrThrow()},
 * et non a un tenant explicite. Le {@code hotelId} n'apparait nulle part en
 * parametre — c'est intentionnel. Un user authentifie n'a pas besoin du
 * tenant pour modifier ses infos perso.</p>
 *
 * <h2>Champs immutables via cet API</h2>
 * <p>{@code username}, {@code email}, {@code role}, {@code hotel}, {@code actif}
 * ne peuvent pas etre modifies via le self-service. Workflow change-email =
 * tour separe (validation par token mail).</p>
 */
public interface ProfileService {

    /**
     * Renvoie le profil complet de l'utilisateur courant. Leve
     * {@link com.cityprojects.citybackend.exception.BusinessException}
     * ({@code error.user.unknown}) si pas d'identite resolue.
     */
    ProfileDto findCurrent();

    /**
     * Applique le DTO PATCH (prenom + nom + telephone + poste). {@code username},
     * {@code email}, {@code role}, {@code hotel}, {@code actif} restent immuables.
     */
    ProfileDto updateCurrent(ProfileUpdateDto dto);

    /**
     * Rotation du mot de passe self-service. Verifie ancien BCrypt, valide
     * nouveau (regex + dictionnaire commun via {@code PasswordUtil}), persiste
     * le nouveau hash, force {@code motPasseTemporaire=false}.
     *
     * @throws com.cityprojects.citybackend.exception.BusinessException si
     *         ancien faux ({@code error.user.password.invalid}), nouveau != confirmation
     *         ({@code error.user.password.mismatch}), nouveau == ancien
     *         ({@code error.user.password.unchanged}), ou faible
     *         ({@code error.user.password.weak}).
     */
    void changePassword(ChangePasswordDto dto);

    /**
     * Upload d'avatar. Validation type + taille via {@link AvatarStorageService}.
     * Renvoie le profil mis a jour (avec {@code avatarUrl} positionne).
     */
    ProfileDto uploadAvatar(MultipartFile file);

    /**
     * Supprime l'avatar courant (fichier disque + clear de la colonne BDD).
     * Idempotent : ne leve pas si pas d'avatar.
     */
    ProfileDto deleteAvatar();
}

File: cityfrontend/src/app/services/profile.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';

import { environment } from '../../environments/environment';
import { ApiResponse } from '../shared/models/api.model';
import { ChangePasswordRequest, Profile, ProfileUpdate } from './profile.model';

/**
 * Service HTTP — Profil self-service de l'utilisateur courant.
 *
 * Endpoints (cf. Tour A backend, 2026-05-13) :
 *   - GET    /api/profile/me                    → ProfileDto
 *   - PUT    /api/profile/me                    → ProfileDto (body: ProfileUpdateDto)
 *   - POST   /api/profile/me/change-password    → void (body: ChangePasswordDto)
 *   - POST   /api/profile/me/avatar             → ProfileDto (multipart "file")
 *   - DELETE /api/profile/me/avatar             → ProfileDto
 *
 * Le backend résout TOUJOURS l'userId courant via SecurityContextHolder
 * (cf. `SecurityUtils.currentUserIdOrThrow()`). Aucun userId / hotelId
 * n'est envoyé par le client — la garde anti-spoof multi-tenant est
 * assurée côté serveur.
 */
@Injectable({ providedIn: 'root' })
export class ProfileService {
  private readonly base = `${environment.apiUrl}/api/profile/me`;

  constructor(private readonly http: HttpClient) {}

  /** Charge le profil courant. */
  getProfile(): Observable<Profile> {
    return this.http
      .get<ApiResponse<Profile>>(this.base)
      .pipe(map((r) => r.data as Profile));
  }

  /** Met à jour les infos personnelles éditables (prenom, nom, telephone, poste). */
  updateProfile(dto: ProfileUpdate): Observable<Profile> {
    return this.http
      .put<ApiResponse<Profile>>(this.base, dto)
      .pipe(map((r) => r.data as Profile));
  }

  /**
   * Change le mot de passe. Le backend valide :
   *  - `ancienMotDePasse` correspond au hash en base.
   *  - `nouveauMotDePasse === confirmation`.
   *  - `nouveauMotDePasse !== ancienMotDePasse`.
   *  - `nouveauMotDePasse` respecte la policy (`PasswordUtil.validatePassword`).
   *
   * Clés d'erreur i18n possibles côté serveur :
   *   - `error.user.password.invalid` (ancien mdp incorrect)
   *   - `error.user.password.mismatch` (confirmation ≠ nouveau)
   *   - `error.user.password.unchanged` (nouveau identique à ancien)
   *   - `error.user.password.weak` (policy non respectée)
   */
  changePassword(dto: ChangePasswordRequest): Observable<void> {
    return this.http
      .post<ApiResponse<void>>(`${this.base}/change-password`, dto)
      .pipe(map(() => undefined));
  }

  /**
   * Upload de l'avatar (multipart). Backend valide :
   *  - taille max 2 MB
   *  - types autorisés : image/jpeg, image/png, image/webp
   *
   * Le composant valide AUSSI côté front avant l'appel pour une meilleure UX.
   */
  uploadAvatar(file: File): Observable<Profile> {
    const form = new FormData();
    form.append('file', file);
    return this.http
      .post<ApiResponse<Profile>>(`${this.base}/avatar`, form)
      .pipe(map((r) => r.data as Profile));
  }

  /** Supprime l'avatar et renvoie le profil mis à jour (avatarUrl = null). */
  deleteAvatar(): Observable<Profile> {
    return this.http
      .delete<ApiResponse<Profile>>(`${this.base}/avatar`)
      .pipe(map((r) => r.data as Profile));
  }

  /**
   * Construit l'URL absolue à utiliser dans un `<img [src]>` pour afficher
   * l'avatar. Retourne `null` si pas d'avatar (le composant affiche alors
   * un fallback initiales).
   *
   * Le backend sert les fichiers statiques sous le contexte `/citybackend`
   * (cf. `application.yml`), donc on concatène avec `environment.apiUrl`
   * qui inclut déjà ce préfixe.
   */
  buildAvatarUrl(profile: Profile | null | undefined): string | null {
    if (!profile?.avatarUrl) {
      return null;
    }
    const path = profile.avatarUrl.startsWith('/')
      ? profile.avatarUrl
      : `/${profile.avatarUrl}`;
    return `${environment.apiUrl}${path}`;
  }
}
